### ISSUE TITLE
Feat: add `prevPermutation`, `nextPermutationBy` and a version of `nextPermutation` that goes back to the first perm?

### ISSUE THREAD

**Issue by gksato** [haskell/vector#499]
When I was authoring #497, I noticed that there is no implementation of `prevPermutation` or `nextPermutationBy`. I also got surprised that `nextPermutation`, unlike in C++, does not go back to the first permutation when the argument holds the last permutation (I understand the reason for the decision, though).  What do you think of adding the functions below? I'm not sure especially about the prime versions of the functions, but it doesn't hurt to record the idea, anyway.

```haskell
module Data.Vector.Generic.Mutable where

-- | Compute the (lexicographically) next permutation of the given vector in-place.
-- Returns False when the input is the last permutation; in this case the vector
-- will be updated to the first permutation, as aligned with the behavior of the C++ function 
-- @std::next_permutation@.
nextPermutation' :: (PrimMonad m, Ord e, MVector v e) => v e -> m Bool

nextPermutationBy :: (PrimMonad m, MVector v e) => (e -> e -> Ordering) -> v e -> m Bool
nextPermutationBy' :: (PrimMonad m, MVector v e) => (e -> e -> Ordering) -> v e -> m Bool
prevPermutation :: (PrimMonad m, Ord e, MVector v e) => v e -> m Bool
prevPermutation' :: (PrimMonad m, Ord e, MVector v e) => v e -> m Bool
prevPermutationBy :: (PrimMonad m, MVector v e) => (e -> e -> Ordering) -> v e -> m Bool
prevPermutationBy' :: (PrimMonad m, MVector v e) => (e -> e -> Ordering) -> v e -> m Bool
```

**Comment by Shimuuar** [comment]
I think adding full complement of `{next,prev}Permutation{,By}` is quite reasonable whenever we have function which works with `Ord` we provide way to supply comparison function as parameter. And adding `prev` variant seems sensible too even if they could be made using `Down`

I'm not sure about primed versions. Whether it's good idea to expand API that much. That's also poor naming prime s usually reserved for indication of strictness

P.S. I didn't get to benchmarking of PR yet. It's pity we don't have way to ask GHC to specialize large functions very aggressively instead of inlining them

**Comment by gksato** [comment]
Thanks for your comment. It looks like we're on the same page. `By` versions are absolute necessity, `prev` versions are sensible, and primed version is... um, questionable. Even if we put necessity aside (which is a big premise!), it's a terrible name. Maybe `nextPermutationGoBack`? It's not like the intension is crystal clear with this name... Oh, `nextPermutationCpp`... just joked but I don't even know if I can forget this as a joke. 

Or can we just add some guide to the doc comment:
```haskell
-- | Compute the (lexicographically) next permutation of the given vector in-place.
-- Returns False when the input is the last permutation; in this case the vector
-- will not be updated, unlike the behavior of the C++ function 
-- @std::next_permutation@.
--
-- If you want to align the behavior with the C++ version, you need to write the
-- following wrapper yourself:
-- > nextPermutationGoBack v = do
-- >   res <- VGM.nextPermutation v
-- >   if res then VGM.reverse v else return ()
-- >   return res
```

Reply to P.S.: Indeed, it's pity. Take your time, it's only a day from the submission of my PR.

**Comment by Shimuuar** [comment]
I think we should first add four `{next,prev}Permutation{,By}`. They should be added in any case and naming is clear. After that we could return whether we want them and how they should be called

**Comment by gksato** [comment]
> I think we should first add four `{next,prev}Permutation{,By}`. They should be added in any case and naming is clear.

I'm working on it in #498.

**Comment by gksato** [comment]
For the sake of readability of this issue itself: we have added `{next,prev}Permutation{,By}` in #498.

During the implementation of #498, I think I have devised a suitable naming convention: `{next,prev}PermutationBijective{,By}`. However, "Bijective" may be unfamiliar to some users.

Regarding the necessity of the primed/"bijective" versions, they have many useful properties for testing and benchmarking `{next,prev}Permutation` themselves. However, I'm uncertain about their value to end users. We could keep this issue open until someone comes in.

**Comment by Shimuuar** [comment]
Another naming possibility is to use `cycle`. It's both shorter and I think clearer but I'm not sure where to add it to identifier name

**Comment by gksato** [comment]
True. If we prioritize grammatical readability, we could add `Cyclically` or `InCycle` at end of the ident, but it's longer (the latter is shorter than `Bijective`, though).
